src/graph/graph.ts

```diff
@@ -398,161 +422,120 @@ export const compileNode = (
+      if (!nodeContext) {
+        return;
+      }
+
+      // Produce the input filler
+      let filler: InputFillerGroup;
+      let fillerName: string | undefined;
+      if (input.property) {
+        fillerName = ensure(
+          (codeNode.config.properties || []).find(
+            (p) => p.property === input.property,
+          )?.fillerName,
+          `Node "${node.name}" has no property named "${input.property}" to find the filler for`,
+        );
+        filler = inputFillers[fillerName];
+      } else {
+        filler = inputFillers[input.id];
+      }
+      if (!filler) {
+        console.error('No filler for property', {
+          input,
+          node,
+          inputFillers,
+          fillerName,
+        });
+        throw new Error(
+          `Node "${node.name}"${
             (node as SourceNode).stage ? ` (${(node as SourceNode).stage})` : ''
-          } has no input ${edge.input}!\nAvailable:${inputs
-            .map(({ id }) => id)
-            .join(', ')}`,
-        ),
-      }))
-      .filter(({ input }) => !isDataInput(input))
-      .forEach(({ fromNode, edge, input }) => {
-        const [inputSections, fillerAst, childIds] = compileNode(
-          engine,
-          graph,
-          edges,
-          engineContext,
-          fromNode,
-          activeIds,
+          } has no filler for input "${input.displayName}" named ${fillerName}`,
         );
-        if (!fillerAst) {
-          throw new TypeError(
-            `Expected a filler ast from node ID ${fromNode.id} (${fromNode.type}) but none was returned`,
-          );
-        }
-
-        continuation = mergeShaderSections(continuation, inputSections);
-        compiledIds = { ...compiledIds, ...childIds };
-
-        let filler: InputFillerGroup;
-        let fillerName: string | undefined;
-        if (nodeContext) {
-          if (input.property) {
-            fillerName = ensure(
-              ((node as CodeNode).config.properties || []).find(
-                (p) => p.property === input.property,
-              )?.fillerName,
-              `Node "${node.name}" has no property named "${input.property}" to find the filler for`,
+      }
+
+      // Test if it needs to be backfilled - this only goes one level deep
+      // because we're only backfilling fromNode
+      let backfillers = codeNode.backfillers?.[input.id];
+      if (backfillers && shouldNodeHaveMainFn(fromNode)) {
+        const childAst = engineContext.nodes[fromNode.id].ast;
+        // For now we can only backfill programs
+        if (childAst.type === 'program') {
+          backfillers.forEach((backfiller) => {
+            // This is where the variable name gets injected into the main
+            // function parameter of the backfilled child
+            backfillAst(
+              childAst,
+              backfiller.argType,
+              backfiller.targetVariable,
+              engineContext.nodes[fromNode.id].mainFn,
```

Comment:
this is the second most important pat of this PR: This mutates the shader module to inject the backfilled variable into its main function is a parameter.

As an example, you have a shader node in the graph, which might have the source code
```glsl
attribute vec2 vUv;

void main() {
   vec4 color = texture2D(image, vUv);
}
```

After backfilling, this becomes

```glsl
attribute vec2 vUv;

void main(vec2 vUv) {
   vec4 color = texture2D(image, vUv);
}
```

It takes advantage of scope rules inherent in GLSL to override the global `vUv` variable with a local one.